src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs

```diff
@@ -230,23 +279,32 @@ public virtual string CreateToken(string payload, SigningCredentials signingCred
             return CreateTokenPrivate(JObject.Parse(payload), signingCredentials, encryptingCredentials, compressionAlgorithm);
         }
 
-        private string CreateTokenPrivate(JObject payload, SigningCredentials signingCredentials, EncryptingCredentials encryptingCredentials, string algorithm)
+        private string CreateTokenPrivate(JObject payload, SigningCredentials signingCredentials, EncryptingCredentials encryptingCredentials, string compressionAlgorithm)
         {
-            if (!JsonWebTokenManager.KeyToHeaderCache.TryGetValue(JsonWebTokenManager.GetHeaderCacheKey(signingCredentials), out string rawHeader))
+            var rawHeader = string.Empty;
+            if (signingCredentials != null && !JsonWebTokenManager.KeyToHeaderCache.TryGetValue(JsonWebTokenManager.GetHeaderCacheKey(signingCredentials), out rawHeader))
             {
                 var header = new JObject
-                {
-                    { JwtHeaderParameterNames.Alg, signingCredentials.Algorithm },
-                    { JwtHeaderParameterNames.Kid, signingCredentials.Key.KeyId },
-                    { JwtHeaderParameterNames.Typ, JwtConstants.HeaderType }
-                };
+                    {
+                        { JwtHeaderParameterNames.Alg, signingCredentials.Algorithm },
+                        { JwtHeaderParameterNames.Kid, signingCredentials.Key.KeyId },
+                        { JwtHeaderParameterNames.Typ, JwtConstants.HeaderType }
+                    };
 
                 if (signingCredentials.Key is X509SecurityKey x509SecurityKey)
                     header[JwtHeaderParameterNames.X5t] = x509SecurityKey.X5t;
 
                 rawHeader = Base64UrlEncoder.Encode(Encoding.UTF8.GetBytes(header.ToString(Formatting.None)));
                 JsonWebTokenManager.KeyToHeaderCache.TryAdd(JsonWebTokenManager.GetHeaderCacheKey(signingCredentials), rawHeader);
             }
+            else if (signingCredentials == null)
+            {
+                var header = new JObject
+                {
+                    { JwtHeaderParameterNames.Alg, SecurityAlgorithms.None }
+                };
+                rawHeader = Base64UrlEncoder.Encode(Encoding.UTF8.GetBytes(header.ToString(Formatting.None)));
```

Comment:
this could be a static constant, create once.

Reply:
Good point! Just made this change. :)